Add Windows installer script and workflow

Wengier opened the pull request:
This adds Inno-based installation script and workflow to automate the building of Windows installer containing 64-bit MSYS2 build by default, which one click switch to 64-bit MSVC build if desired. Desktop icons and shell extensions are installed by default, although both can be optionally disabled by users via checkboxes.

Kappa971 commented:
Does this installer also add the "open with DOSBox" option to the Windows context menu?
If so, has the new Windows 11 context menu also been taken into consideration? Currently, I only know of WinRar and WinMerge which support the new context menu, but that makes Windows 11 more usable.

If it has nothing to do with this PR, well, I haven't written anything 😁

Wengier commented:
@Kappa971 Yes, that is what "shell extensions" refer to in my original post. It works in Windows 11 too, although in this case you need to click "Show more options" and then "Run with DOSBox Staging". You don't need to click "Show more options" in Windows 10 (or earlier).

kcgen commented:
For the logo, we've got these existing resources in PNG (small, medium, large), and as vector-graphics:

```
./src/dosbox_staging_splash_640_400.png
./src/dosbox_staging_splash.svg
./contrib/fedora/dosbox-staging.spec
./contrib/linux/dosbox-staging.metainfo.xml
./contrib/linux/dosbox-staging.desktop
./contrib/icons/dosbox-staging.svg
./contrib/icons/small-svg/dosbox-staging-32.svg
./contrib/icons/small-svg/dosbox-staging-24.svg
./contrib/icons/small-svg/dosbox-staging-16.svg
./contrib/icons/small-svg/dosbox-staging-src.svg
./contrib/icons/dosbox-staging.ico
./contrib/icons/small-png/16x16/dosbox-staging.png
./contrib/icons/small-png/32x32/dosbox-staging.png
./contrib/icons/small-png/22x22/dosbox-staging.png
./contrib/icons/small-png/24x24/dosbox-staging.png
```

If it's possible to leverage any of them (and use a relative path, like `../icons/`, that would be ideal, as it's fewer copies.

But no worries, if it can't use those format.

kcgen commented:
It would be nice to avoid embedding the `7zr.exe` executable and use what's on the system. 7zip continues to evolve, and this approach also means we don't have to keep updating it (for security fixes, etc).

On the GitHub runners, 7zip is available under:

 - `C:/Program Files/7-Zip`, and also
 - `C:/ProgramData/Chocolatey/bin`

We can also installing it under MSYS2 (as another route to get it), by adding `p7zip` to:

 - `packages/windows-msys2-clang-x86_64.txt`, and
 - `packages/windows-msys2-gcc-x86_64.txt`

And it will automatically be pulled in during the builds, even on local machines.

Kappa971 commented:
> @Kappa971 Yes, that is what "shell extensions" refer to in my original post. It works in Windows 11 too, although in this case you need to click "Show more options" and then "Run with DOSBox Staging". You don't need to click "Show more options" in Windows 10 (or earlier).

Unfortunately, as already mentioned, it is not very comfortable. It would be nice if the new context menu was also supported in Windows 11 without having to click on "Show more options". WinRar for example does this, obviously I don't know how complex it is to do this.

Wengier commented:
@Kappa971 Ideally even this can be solved for better Windows 11 support, but unfortunately it does not appear to be so easy to find this technical information yet (if one knows how this can be done please let me know), and the shell extension is considered an add-on function for the installer. But if one feels the new context menu is annoying, they really do not need to wait for Microsoft to go back to the old menus, since it appears to be quite easy to find information online detailing how to bring back the "old" context menu on Windows 11, e.g.:

https://pureinfotech.com/bring-back-classic-context-menu-windows-11/

Wengier commented:
@kcgen Thanks for the information about 7zip! The built-in 7z.exe is now called instead of 7zr.exe. As for PNG and SVG, they are unfortunately not supported by Inno Setup, but I have now moved dosbox-staging.bmp to ./contrib/icons (same as other images). I have also removed the setup-preamble.txt file and let the installer use the docs/README.template instead.

Kappa971 commented:
> @Kappa971 Ideally even this can be solved for better Windows 11 support, but unfortunately it does not appear to be so easy to find this technical information yet (if one knows how this can be done via registry or programmatically please let me know), and the shell extension is considered an (optional) add-on function for the installer. But if one feels the new context menu is annoying, they really do not need to wait for Microsoft to go back to the old menus, since it appears to be quite easy to find information online detailing how to bring back the "old" context menu on Windows 11, e.g.:
> 
> https://pureinfotech.com/bring-back-classic-context-menu-windows-11/

This is what I found, I don't know if it can help (unfortunately I can't help you in any other way than with testing): https://blogs.windows.com/windowsdeveloper/2021/07/19/extending-the-context-menu-and-share-dialog-in-windows-11/

Here are the WinMerge commits (marked with "Shell Extension for Windows 11 or later"), maybe it's not useful for DOSBox but I'll put the link anyway: https://github.com/WinMerge/winmerge/commits/master?after=73362f0c57e4a94522840750be60daff44089d36+594&branch=master&qualified_name=refs%2Fheads%2Fmaster

Grounded0 commented:
It's good enough as is for 0.79 for now. Enhancements can come later as Windows 11 adoption picks up. Shipping should be utmost priority now.

kcgen commented:
@Wengier , 

I'm looking at PR's MSYS2 and Windows CI build summary pages:
 - https://github.com/dosbox-staging/dosbox-staging/actions/runs/3035951368
 - https://github.com/dosbox-staging/dosbox-staging/actions/runs/3035951359
 
Can you link to the installers, so Windows users can try them?

kcgen commented:
Direct download for Windows installer: https://github.com/dosbox-staging/dosbox-staging/suites/8242849909/artifacts/360150975

Thanks @Wengier !

Grounded0 commented:
@Burrito78 reports that in German Windows it offers C:\Program Files as default install directory when in German Windows it should be C:\Programme.

"At system startup, the actual paths to 'Program Files' and 'Common Program Files' are loaded from the Windows registry, where they are stored in the ProgramFilesDir and CommonFilesDir values under HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion. They are then made accessible to the rest of the system via the volatile environment variables %ProgramFiles%, and %CommonProgramFiles%. Applications can also obtain the locations of these paths by querying the Setup API using dirids, or through Windows Management Instrumentation, or by querying the shell using CSIDLs, or ShellSpecialFolderConstants. These are all localization-independent methods."

Using %ProgramFiles% or similar idiom should offer install to localised version of "Program Files" to every user in the world:
https://en.wikipedia.org/wiki/Program_Files

kcgen commented on contrib/windows_installer/DOSBox-Staging-setup.iss:
```diff
@@ -0,0 +1,130 @@
+Root: HKA; Subkey: "Software\Classes\SystemFileAssociations\.bat\shell\Run with DOSBox Staging\command"; ValueType: none; Check: not WizardIsTaskSelected('contextmenu'); Flags: deletekey
+Root: HKA; Subkey: "Software\Classes\SystemFileAssociations\.conf\shell\Open with DOSBox Staging"; ValueType: none; Check: not WizardIsTaskSelected('contextmenu'); Flags: deletekey
+Root: HKA; Subkey: "Software\Classes\SystemFileAssociations\.conf\shell\Open with DOSBox Staging\command"; ValueType: none; Check: not WizardIsTaskSelected('contextmenu'); Flags: deletekey
+
```
Is it possible to replace most or all the above 'DOSBox Staging' strings with the one definition at the top, `{#DOSBoxAppName}`?

Wengier replied:
Thanks for the suggestion! Done.

kcgen commented on contrib/windows_installer/DOSBox-Staging-setup.iss:
```diff
@@ -0,0 +1,130 @@
+
+[Files]
+Source: "program\*"; DestDir: "{app}"; Flags: ignoreversion recursesubdirs createallsubdirs
+Source: "dosbox.exe"; DestDir: "{app}"; DestName: "dosbox.exe"; Flags: ignoreversion; Tasks: not defaultmsvc
```
Is it possible to peplace all the `dosbox.exe` instances with `{#DOSBoxAppExeName}`.

And make new `define`'s for `dosbox_with_debugger.exe`, `dosbox_msvc.exe`, `dosbox_msvc_with_debugger.exe`, and `dosbox_with_console.bat` -> and swap those in, as well?

Wengier replied:
Done replacing all the instances.

kcgen commented on contrib/windows_installer/DOSBox-Staging-setup.iss:
```diff
@@ -0,0 +1,130 @@
+Name: "{group}\{#DOSBoxAppName}"; Filename: "{app}\dosbox.exe"; Parameters: "-noconsole"
+Name: "{group}\{#DOSBoxAppName} with console"; Filename: "{app}\dosbox_with_console.bat"
+Name: "{group}\{#DOSBoxAppName} with debugger"; Filename: "{app}\dosbox_with_debugger.exe"
+Name: "{group}\Visit {#DOSBoxAppName} website"; Filename: "https://dosbox-staging.github.io/"
```
The https URL can be replaced with the `{#DOSBoxAppURL}` define.

Wengier replied:
Done replacing too.

kcgen commented on contrib/windows_installer/DOSBox-Staging-setup.iss:
```diff
@@ -0,0 +1,130 @@
+Name: "{autodesktop}\{#DOSBoxAppName}"; Filename: "{app}\dosbox.exe"; Parameters: "-noconsole"; Tasks: desktopicon
+
+[Registry]
+Root: HKA; Subkey: "Software\dosbox-staging"; Flags: uninsdeletekeyifempty
+Root: HKA; Subkey: "Software\dosbox-staging"; ValueType: string; ValueName: "Path"; ValueData: "{app}"; Flags: uninsdeletekey
+Root: HKA; Subkey: "Software\dosbox-staging"; ValueType: string; ValueName: "Version"; ValueData: "{#DOSBoxAppVersion}"; Flags: uninsdeletekey
+Root: HKA; Subkey: "Software\Classes\Directory\shell\dosbox-staging"; ValueType: string; ValueName: ""; ValueData: "Open with {#DOSBoxAppName}"; Check: WizardIsTaskSelected('contextmenu'); Flags: uninsdeletevalue
+Root: HKA; Subkey: "Software\Classes\Directory\shell\dosbox-staging"; ValueType: string; ValueName: "Icon"; ValueData: """{app}\dosbox_with_console.bat"",0"; Check: WizardIsTaskSelected('contextmenu'); Flags: uninsdeletekey
+Root: HKA; Subkey: "Software\Classes\Directory\shell\dosbox-staging\command"; ValueType: string; ValueName: ""; ValueData: """{app}\dosbox_with_console.bat"" ""%v"""; Check: WizardIsTaskSelected('contextmenu'); Flags: uninsdeletekey
+Root: HKA; Subkey: "Software\Classes\Directory\Background\shell\dosbox-staging"; ValueType: string; ValueName: ""; ValueData: "Open with {#DOSBoxAppName}"; Check: WizardIsTaskSelected('contextmenu'); Flags: uninsdeletevalue
+Root: HKA; Subkey: "Software\Classes\Directory\Background\shell\dosbox-staging"; ValueType: string; ValueName: "Icon"; ValueData: """{app}\dosbox_with_console.bat"",0"; Check: WizardIsTaskSelected('contextmenu'); Flags: uninsdeletekey
+Root: HKA; Subkey: "Software\Classes\Directory\Background\shell\dosbox-staging\command"; ValueType: string; ValueName: ""; ValueData: """{app}\dosbox_with_console.bat"" ""%v"""; Check: WizardIsTaskSelected('contextmenu'); Flags: uninsdeletekey
```
there are lots of repeating `dosbox-staging` strings - can you make a `define` for them, and swap it in for all of them too?

Wengier commented:
Thanks for the reviews. For a fresh installation the installer will now show a dialog for selecting the install mode, including installations for the current user (installing to the user program directory) or for all users (installing to %ProgramFiles%), and the former is actually recommended for many cases, as also mentioned in the "Select install mode" dialog itself.

Wengier commented:
Merged this, and Windows installers will be automatically generated by the CI for all future commits from now on, downloadable as an artifact in the "MSYS2 builds" workflow.
